js/src/main/scala/org/soundsofscala/instrument/SamplePlayer.scala

```diff
@@ -22,15 +22,30 @@ import org.scalajs.dom
 import org.scalajs.dom.AudioBuffer
 import org.scalajs.dom.AudioContext
 import org.soundsofscala.models.AtomicMusicalEvent
+import scala.scalajs.js.typedarray.Float32Array
+import scala.scalajs.js.JSConverters.iterableOnceConvertible2JSRichIterableOnce
 
-trait SamplePlayer extends Instrument {}
+trait SamplePlayer extends Instrument[SamplePlayer.Settings] {}
 
 object SamplePlayer {
+  final case class Settings(
+      playbackRate: Double,
+      reversed: Boolean,
+      loop: Boolean,
```

Comment:
`loopStart` and `loopEnd` don't have any meaning unless `loop` is true, correct? In that case this is better modelled as a sum type. The following is one way to do it.

```scala
final case class Loop(start: Int, end: Int)

final case class Settings(
  ...
  loop: Option[Loop]
)
```

Reply:
I thought of another issue: what if the end is before the start? Is this an error, or does this mean a loop is reversed? If it means a loop is reversed that means the `reversed` flag is not needed and there isn't the need to check that `start` < `end`.

Reply:
I checked the web audio api docs and it doesn't reverse, it just loops over the whole track if loopEnd is < or = to loopStart. So no issue there except for informing the user that that's the case?